### PR TITLE
Add group-based and owner-based query access rules

### LINKED ISSUE #9484: Missing group attribute for queries in File system access control
Hi Trino-Team!

Reading and trying the docs https://trino.io/docs/current/security/file-system-access-control.html#query-rules i noticed two issues:
1. Bug: The attribute "owner" does not get pick up by the JSON Mapper [QueryAccessRule](https://github.com/trinodb/trino/blob/caf141718b79948c21764e76fa91249c3e1ade46/lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/QueryAccessRule.java#L40). I also don't see any usage of this e.g. in [FileBasedSystemAccessControl](https://github.com/trinodb/trino/blob/master/lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/FileBasedSystemAccessControl.java). IMHO this is a bug, because users limiting user A to kill queries of user B would instead grant user A to kill any query regardless of the owner.
	a. We can either remove the "owner" attribute from the documentation.
	b. I can think of some use-cases where user A should be able to stand in for user B and kill his query
I would suggest sticking with option b as the interface [SystemAccessControl](https://github.com/trinodb/trino/blob/master/core/trino-spi/src/main/java/io/trino/spi/security/SystemAccessControl.java) passes the owner attribute e.g. `default void checkCanKillQueryOwnedBy(SystemSecurityContext context, String queryOwner)`
2. Missing feature: We are not able to give the group "admin" the ability to kill all queries. The attribute group is missing, we can only grant it to explicit users. This is already implemented for roles in [QueryAccessRule](https://github.com/trinodb/trino/blob/caf141718b79948c21764e76fa91249c3e1ade46/lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/QueryAccessRule.java#L50) so i should be possible to easily implement this

Regards,
Sebastian

### PR DESCRIPTION
**Pull request by cccs-tom**
Fixes #9484.
QueryAccessRule now supports a `group` regex to enable rules based on group membership and an `owner` regex to enable rules based on the query's owner.

### CODE REVIEW

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/QueryAccessRule.java [hunk 1]
```diff
@@ -35,22 +35,30 @@
...
+    private final Optional<Pattern> ownerRegex;
 
     @JsonCreator
     public QueryAccessRule(
             @JsonProperty("allow") Set<AccessMode> allow,
             @JsonProperty("user") Optional<Pattern> userRegex,
-            @JsonProperty("role") Optional<Pattern> roleRegex)
+            @JsonProperty("role") Optional<Pattern> roleRegex,
+            @JsonProperty("group") Optional<Pattern> groupRegex,
+            @JsonProperty("owner") Optional<Pattern> ownerRegex)
```
[the remark this answers is not in the document]
Comment (kokosing) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/QueryAccessRule.java:48]: > If you still feel that the owner-based rules belong in a separate PR, I can split it. Please let me know.

I think it is fine. However I would make this PR to have 2 commits. Notice the PR title `Add group-based and owner-based query access rules`. It is natural to split this using "delimiter" `and` into:
 - `Add group-based query file based access rules`.
 - `Add owner-based query file based access rules`.

**Review by kokosing — COMMENTED** [review]
Looks nice!

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/QueryAccessRule.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (cccs-tom) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/QueryAccessRule.java:48]: No problem, I can do that.

File: lib/trino-plugin-toolkit/src/test/resources/query-invalid.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+  "queries": [
+    {
+      "group": "contractors",
+      "owner": "alice|dave",
```
[the remark this answers is not in the document]
Comment (cccs-tom) [reply lib/trino-plugin-toolkit/src/test/resources/query-invalid.json:5]: I have no objections. I do want to note however that the `owner` field was already in the documentation - it was simply ignored by the code. If you're ok with changing it, then so am I! :)

You prefer `queryOwnedBy` over `queryOwner`? I don't mind either way - just checking.

I'll be honest, I'm assuming it's just a matter of time for someone to request support for `queryOwnerMemberOf` (for groups) and `queryOwnerHasRole`...

[the remark this answers is not in the document]
Comment (cccs-tom) [reply lib/trino-plugin-toolkit/src/test/resources/query-invalid.json]: Ok, I've split the commits and renamed the `owner` field to `onQueryOwnedBy`, as discussed.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 3]
```diff
@@ -339,16 +339,22 @@ owned by any user. If no rule matches, query management is denied. Each rule is
 of the following fields:
 
 * ``user`` (optional): regex to match against user name. Defaults to ``.*``.
-* ``owner`` (optional): regex to match against the query owner name. Defaults to ``.*``.
+* ``role`` (optional): regex to match against role names. Defaults to ``.*``.
+* ``group`` (optional): regex to match against group names. Defaults to ``.*``.
+* ``onQueryOwnedBy`` (optional): regex to match against the query owner name. Defaults to ``.*``.
```
Comment (kokosing) [inline docs/src/main/sphinx/security/file-system-access-control.rst:344]: Let's use `queryOwner`

**Review by kokosing — APPROVED** [review]

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 3 above)
Comment (cccs-tom) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Done

**Review by kokosing — APPROVED** [review]

**Comment by kokosing** [comment]
Merged, thanks!

**Comment by sbernauer** [comment]
Thanks a lot for working on this @cccs-tom!

**Pull request merged**